packages/core/src/hydration/skip_hydration.ts

```diff
@@ -0,0 +1,77 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {assertTNodeForLView} from '../render3/assert';
+import {getTNodeFromLView} from '../render3/di';
+import {TContainerNode, TElementNode, TNode} from '../render3/interfaces/node';
+import {DECLARATION_VIEW, LView} from '../render3/interfaces/view';
+
+export const SKIP_HYDRATION_ATTR_NAME = 'ngSkipHydration';
+
+/**
+ * Helper function to walk up parent nodes using TNode data structure, crossing
+ * view boundaries if needed, calling `predicateFn` at each level (with the current
+ * TNode as an argument). The process stops when predicate return `true` for
+ * the first time. If `predicateFn` never returned `true` after reaching the root
+ * view, the function returns `false`.
+ *
+ * @param tNode
+ * @param lView
+ * @param predicateFn
+ * @returns
+ */
+export function navigateParentTNodes(
+    tNode: TNode, lView: LView, predicateFn: (tNode: TNode) => boolean): TNode|null {
+  let currentTNode: TNode|null = tNode;
+  let currentLView: LView|null = lView;
+
+  while (currentTNode !== null && currentLView !== null) {
+    ngDevMode && assertTNodeForLView(currentTNode, currentLView);
+
+    if (predicateFn(currentTNode)) {
+      return currentTNode;
+    }
+
+    // Has an explicit type due to a TS bug: https://github.com/microsoft/TypeScript/issues/33191
+    let parentTNode: TElementNode|TContainerNode|null = currentTNode.parent;
+
+    // `TNode.parent` includes the parent within the current view only. If it doesn't exist,
+    // it means that we've hit the view boundary and we need to go up to the next view.
+    if (!parentTNode) {
+      // Keep going up the tree.
+      parentTNode = getTNodeFromLView(currentLView);
+      currentLView = currentLView[DECLARATION_VIEW];
+    }
+
+    currentTNode = parentTNode;
+  }
+
+  return null;
+}
+
+export function hasNgSkipHydrationAttr(tNode: TNode): boolean {
+  const attrs = tNode.mergedAttrs;
+  if (attrs === null) return false;
+  // only ever look at the attribute name and skip the values
+  for (let i = 0; i < attrs.length; i += 2) {
+    const value = attrs[i];
+    if (typeof value === 'string' &&
+        (value as string).toLowerCase() === SKIP_HYDRATION_ATTR_NAME.toLowerCase()) {
+      return true;
+    }
+  }
+  return false;
+}
+
+export function isInSkipHydrationBlock(tNode: TNode, lView: LView): boolean {
+  const foundTNode = navigateParentTNodes(tNode as TNode, lView, hasNgSkipHydrationAttr);
+  // We are in a skip hydration block when:
+  // - we have a TNode
+  // - the tNode is different than the root node
+  return foundTNode !== null && foundTNode !== tNode;
```

Comment:
(we can also rename `foundTNode` -> `skipHydrationRootTNode`)
```suggestion
  // We are in a skip hydration block when:
  // - a tNode holding the `ngSkipHydration` attribute is found
  // - the found tNode is different than the one that we've started with
  return skipHydrationRootTNode !== null && skipHydrationRootTNode !== tNode;
```